app/code/community/Bolt/Boltpay/Helper/Coupon.php

```diff
@@ -0,0 +1,662 @@
+        $immutableQuote = $this->getImmutableQuote();
+        // rule per customer usage
+        if ($usesPerCustomer = $rule->getUsesPerCustomer()) {
+            $customerId = $immutableQuote->getCustomerId();
+            $ruleCustomer = Mage::getModel('salesrule/rule_customer')->loadByCustomerRule($customerId, $rule->getId());
+            if ($ruleCustomer->getId() && $ruleCustomer->getTimesUsed() >= $usesPerCustomer) {
+                $this->setErrorResponseAndThrowException(
+                    self::ERR_CODE_LIMIT_REACHED,
+                    sprintf('The code [%s] has exceeded usage limit.', $this->getCouponCode()),
+                    422,
+                    $immutableQuote
+                );
+            }
+        }
+    }
+
+    /**
+     * Attempts to apply the coupon to both the immutable and session quote.
+     */
+    protected function applyCouponToQuotes()
+    {
+        try {
+
+            $parentQuote = $this->getParentQuote();
+            $immutableQuote = $this->getImmutableQuote();
+            $couponCode = $this->getCouponCode();
+
+            // Try applying to parent first
+            $this->applyCouponToQuote($parentQuote, $couponCode);
+            // Apply coupon to clone
+            $this->applyCouponToQuote($immutableQuote, $couponCode);
+        } catch (\Exception $e) {
+            $this->setErrorResponseAndThrowException(
+                self::ERR_SERVICE,
+                $e->getMessage(),
+                422,
+                @$immutableQuote
+            );
+        }
+    }
+
+    /**
+     * Applies the coupon code to the quote and recalculates totals.
+     *
+     * @param Mage_Sales_Model_Quote $quote
+     * @param string $couponCode
+     */
+    protected function applyCouponToQuote($quote, $couponCode)
+    {
+        $quote->getShippingAddress()->setCollectShippingRates(true);
+        $quote->setCouponCode($couponCode);
+
+        Mage::helper('boltpay')->collectTotals($quote, true)->save();
+    }
+
+    /**
+     * Verifies that the coupon code was applied to the quote.
+     *
+     * @throws Exception     thrown if coupon in quote doesn't match the coupon applied.
+     */
+    protected function validateAfterApplyingCoupon()
+    {
+        $immutableQuote = $this->getImmutableQuote();
+        if ($immutableQuote->getCouponCode() != $this->getCouponCode()) {
+            $this->setErrorResponseAndThrowException(
+                self::ERR_SERVICE,
+                sprintf("Invalid coupon code response for coupon [%s]", $this->getCouponCode()),
+                422,
+                $immutableQuote
+            );
+        }
+    }
+
+    /**
+     * Sets the response error and the http status code and then throws an exception.
+     * @param int $errCode
+     * @param string $message
+     * @param int $httpStatusCode
+     * @throws Exception
+     */
+    protected function setErrorResponseAndThrowException($errCode, $message, $httpStatusCode)
```

Comment:
I think this method should have optional 4th argument Exception and chain it to `throw new \Exception($message)` below